mongo/client_encryption.go

```diff
@@ -142,6 +204,175 @@ func (ce *ClientEncryption) Close(ctx context.Context) error {
 	return ce.keyVaultClient.Disconnect(ctx)
 }
 
+// DeleteKey removes the key document with the given UUID (BSON binary subtype 0x04) from the key vault collection.
+// Returns the result of the internal deleteOne() operation on the key vault collection.
+func (ce *ClientEncryption) DeleteKey(ctx context.Context, id primitive.Binary) (*DeleteResult, error) {
+	return executeDeleteResult(ce, &id, func(coll *Collection, query []byte) (*DeleteResult, error) {
+		return coll.DeleteOne(ctx, query)
+	})
+}
+
+// GetKeyByAltName returns a key document in the key vault collection with the given keyAltName.
+func (ce *ClientEncryption) GetKeyByAltName(ctx context.Context, keyAltName string) *SingleResult {
+	return executeSingleResult(ce, nil, func(coll *Collection, _ []byte) *SingleResult {
+		filter := bsoncore.NewDocumentBuilder().AppendString("keyAltNames", keyAltName).Build()
+		return coll.FindOne(ctx, filter)
+	})
+}
+
+// GetKey finds a single key document with the given UUID (BSON binary subtype 0x04). Returns the result of the
+// internal find() operation on the key vault collection.
+func (ce *ClientEncryption) GetKey(ctx context.Context, id primitive.Binary) *SingleResult {
+	return executeSingleResult(ce, &id, func(coll *Collection, query []byte) *SingleResult {
+		return coll.FindOne(ctx, query)
+	})
+}
+
+// GetKeys finds all documents in the key vault collection. Returns the result of the internal find() operation on the
+// key vault collection.
+func (ce *ClientEncryption) GetKeys(ctx context.Context) (*Cursor, error) {
+	coll, err := ce.majorityReadCollection()
+	if err != nil {
+		return newEmptyCursor(), nil
+	}
+	return coll.Find(ctx, bson.D{})
+}
+
+// RemoveKeyAltName removes a keyAltName from the keyAltNames array of the key document in the key vault collection with
+// the given UUID (BSON binary subtype 0x04). Returns the previous version of the key document.
+func (ce *ClientEncryption) RemoveKeyAltName(ctx context.Context, id primitive.Binary, keyAltName string) *SingleResult {
+	return executeSingleResult(ce, &id, func(coll *Collection, query []byte) *SingleResult {
+		update := bson.A{bson.D{{"$set", bson.D{{"keyAltNames", bson.D{{"$cond", bson.A{bson.D{{"$eq",
+			bson.A{"$keyAltNames", bson.A{keyAltName}}}}, "$$REMOVE", bson.D{{"$filter",
+			bson.D{{"input", "$keyAltNames"}, {"cond", bson.D{{"$ne", bson.A{"$$this", keyAltName}}}}}}}}}}}}}}}
```

Comment:
This is a pretty complicated and unusual query for our code base, but it is 1-1 with the solution in the C driver and, ofc, the unified test expectations. Do we want to reorganize the structure of this code to read better or are we comfortable with the current formatting?  The formatting suggestions from compass language export looks like the following, which seemed a little too heavy for inline use:

```go
bson.A{
    bson.D{
        {"$set",
            bson.D{
                {"keyAltNames",
                    bson.D{
                        {"$cond",
                            bson.A{
                                bson.D{
                                    {"$eq",
                                        bson.A{
                                            "$keyAltNames",
                                            bson.A{
                                                "keyaltname",
                                            },
                                        },
                                    },
                                },
                                "$$REMOVE",
                                bson.D{
                                    {"$filter",
                                        bson.D{
                                            {"input", "$keyAltNames"},
                                            {"cond",
                                                bson.D{
                                                    {"$ne",
                                                        bson.A{
                                                            "$$this",
                                                            "keyaltname",
                                                        },
                                                    },
                                                },
                                            },
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
            },
        },
    },
}
```

Reply:
The formatting you have now seems better to me for readability. Could you point to where this is done in the C driver?

Reply:
Absolutely, [here is the C driver implementation](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/src/mongoc/mongoc-client-side-encryption.c#L2364).

Reply:
The only complaint I have is that this query requires marshaling. Using `bsoncore` as you did for other queries in this file would be good, but I think would be horribly unreadable. I'm ok accepting the tradeoff here, but just noting.

Reply:
If using `bsoncore` helps with readability, that seems preferable to me. I am OK with the proposed use of `bson.D`. The command sent is tested with command started events in specification tests.